api/samples/memval_simple.c

```diff
@@ -0,0 +1,434 @@
+typedef struct {
+    file_t     log;
+    FILE      *logf;
+    reg_id_t   reg_addr;
+} per_thread_t;
+
+static client_id_t client_id;
+static int         tls_idx;
+static drx_buf_t  *write_buffer;
+static drx_buf_t  *trace_buffer;
+
+/* Requires that hex_buf be at least as long as 2*memref->size + 1. */
+static char *
+write_hexdump(char *hex_buf, byte *write_base, mem_ref_t *mem_ref)
+{
+    int i;
+    char *hexstring = hex_buf, *needle = hex_buf;
+
+    for (i = mem_ref->size - 1; i >= 0; --i)
+        needle += dr_snprintf(needle, 2*mem_ref->size+1-2*i, "%02x", write_base[i]);
+    return hexstring;
+}
+
+/* Called when the trace buffer has filled up, and needs to be flushed to disk. */
+static void
+trace_fault(void *drcontext, void *buf_base, size_t size)
+{
+    per_thread_t *data = drmgr_get_tls_field(drcontext, tls_idx);
+    mem_ref_t *trace_base = (mem_ref_t *)(char *)buf_base;
+    mem_ref_t *trace_ptr  = (mem_ref_t *)((char *)buf_base + size);
+    byte *write_base = drx_buf_get_buffer_base(drcontext, write_buffer);
+    byte *write_ptr  = drx_buf_get_buffer_ptr (drcontext, write_buffer);
+    int largest_size = 0;
+    mem_ref_t *mem_ref;
+    char *hex_buf;
+
+    /* find the largest necessary buffer so we only perform a single allocation */
+    for (mem_ref = trace_base; mem_ref < trace_ptr; mem_ref++) {
+        if (mem_ref->size > largest_size)
+            largest_size = mem_ref->size;
+    }
+    hex_buf = dr_thread_alloc(drcontext, 2*largest_size+1);
+    /* write the memrefs to disk */
+    for (mem_ref = trace_base; mem_ref < trace_ptr; mem_ref++) {
+        /* Each memref in the trace buffer has an "associated" write in the write buffer.
+         * We pull mem_reg->size bytes from the write buffer, and assert we haven't yet
+         * gone too far.
+         */
+        /* We use libc's fprintf as it is buffered and much faster than dr_fprintf for
+         * repeated printing that dominates performance, as the printing does here. Note
+         * that a binary dump is *much* faster than fprintf still.
+         */
+        fprintf(data->logf, ""PFX": %s %2d %s\n",
+                (ptr_uint_t)mem_ref->addr, decode_opcode_name(mem_ref->type),
+                mem_ref->size, write_hexdump(hex_buf, write_base, mem_ref));
+        write_base += mem_ref->size;
+        DR_ASSERT(write_base <= write_ptr);
+    }
+    dr_thread_free(drcontext, hex_buf, 2*largest_size+1);
+    /* reset the write buffer (note: the trace buffer gets reset automatically) */
+    drx_buf_set_buffer_ptr(drcontext, write_buffer,
+                           drx_buf_get_buffer_base(drcontext, write_buffer));
+}
+
+static reg_id_t
+instrument_mem(void *drcontext, instrlist_t *ilist, instr_t *where, opnd_t ref)
+{
+    reg_id_t reg_ptr, reg_tmp, reg_addr;
+    ushort type, size;
+    bool ok;
+
+    if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_tmp)
+            != DRREG_SUCCESS) {
```

Comment:
style: indent should be under 'd' of 'drreg' -- but prob fits on line above?  I wish github let me set a max line width and show a marker there.